### PR TITLE
[A2-771] AuthorizedProjects handles deny

### PR DESCRIPTION
**Pull request by bcmdarroch**
### :nut_and_bolt: Description
This PR refines the rego behind ProjectsAuthorized to ensure a deny overrides an allow on the same project, even with statements containing `All-Projects`

### :+1: Definition of Done

- [x] when request includes projects and statement denies all projects, ProjectsAuthorized returns no projects
- [x] when request includes no projects and statement denies all projects, ProjectsAuthorized returns no projects
- [x] when request includes no projects and one statement allows all projects and another statement denies one project, authz-service returns a list of all projects minus the denied projects

### :athletic_shoe: Demo Script / Repro Steps

1. rebuild services  and make some projects, teams, and roles
```
rebuild components/authz-service

export TOK=`chef-automate iam token create ADMIN --admin`
curl -kH "api-token: $TOK" -d '{"name":"project 1", "id": "p1"}' https://a2-dev.test/apis/iam/v2beta/projects | jq
curl -kH "api-token: $TOK" -d '{"name":"project 2", "id": "p2"}' https://a2-dev.test/apis/iam/v2beta/projects | jq

curl -kH "api-token: $TOK" -d '{"name":"role 1", "id": "r1", "actions": ["do:the:thing"], "projects": ["p1"]}' https://a2-dev.test/apis/iam/v2beta/roles | jq
curl -kH "api-token: $TOK" -d '{"name":"role 2", "id": "r2", "actions": ["do:the:thing"], "projects": ["p2"]}' https://a2-dev.test/apis/iam/v2beta/roles | jq

curl -kH "api-token: $TOK" -d '{"name":"team 2", "id": "t2", "projects": ["p2"]}' https://a2-dev.test/apis/iam/v2beta/teams | jq
```

2. upgrade to 2.1 and make a test token
```
chef-automate iam upgrade-to-v2 --beta2.1
export TEST=`chef-automate iam token create test`
```

3. make a policy for that token
```
# save as policy.json
{
  "name": "testing deny",
  "id": "test",
  "members": [ "token:test" ],
  "statements": [
    {
      "effect": "DENY",
      "actions": [ "iam:teams:list" ],
      "projects": [ "*" ]
    },
   {
      "effect": "ALLOW",
      "actions": [ "iam:teams:list" ],
      "projects": [ "p2" ]
    },
  {
      "effect": "ALLOW",
      "actions": [ "iam:roles:list" ],
      "projects": [ "*" ]
    },
   {
      "effect": "DENY",
      "actions": [ "iam:roles:list" ],
      "projects": [ "p2" ]
    }
 ]
}

$ curl -kH "api-token: $TOK" -d @policy.json https://a2-dev.test/apis/iam/v2beta/policies | jq
```

4. run test cases

a. When a user wants a list of teams belonging to `p2`, but a statement denies listing teams for all projects, ProjectsAuthorized returns no projects.
```
$ curl -kH "api-token: $TEST" -H "projects: p2" https://a2-dev.test/apis/iam/v2beta/teams | jq
{
  "error": "unauthorized: subjects [\"token:test\"] cannot perform action \"iam:teams:list\" on resource \"iam:teams\"",
  "message": "unauthorized: subjects [\"token:test\"] cannot perform action \"iam:teams:list\" on resource \"iam:teams\"",
  "code": 7,
  "details": []
}

$ sl | grep authz-service
[2019-04-23T18:39:03+00:00] authz-service.default(O): time="2019-04-23T18:39:03Z" level=info msg="Projects Authorized Query" action="iam:teams:list" projects="[p2]" resource="iam:teams" result="[]" subject="[token:test]"
```

b. When a user wants a list of teams they can see and doesn't apply a project filter, but a statement denies listing teams for all projects, ProjectsAuthorized returns no projects
```
$ curl -kH "api-token: $TEST" https://a2-dev.test/apis/iam/v2beta/teams | jq
{
  "error": "unauthorized: subjects [\"token:test\"] cannot perform action \"iam:teams:list\" on resource \"iam:teams\"",
  "message": "unauthorized: subjects [\"token:test\"] cannot perform action \"iam:teams:list\" on resource \"iam:teams\"",
  "code": 7,
  "details": []
}

$ sl | grep authz-service
[2019-04-23T18:36:54+00:00] authz-service.default(O): time="2019-04-23T18:36:54Z" level=info msg="Projects Authorized Query" action="iam:teams:list" projects="[]" resource="iam:teams" result="[]" subject="[token:test]"
```

c. when request includes no projects and one statement allows all projects and another statement denies one project, authz-service returns a list of all projects minus the denied projects
```
$ curl -kH "api-token: $TEST" https://a2-dev.test/apis/iam/v2beta/roles | jq
# returns chef-managed roles and role 1, no role 2
{
  "roles": [
    {
      "name": "Editor",
      "id": "editor",
      "type": "CHEF_MANAGED",
      "actions": [
        ...
      ],
      "projects": []
    },
    {
      "name": "Owner",
      "id": "owner",
      "type": "CHEF_MANAGED",
      "actions": [
        "*"
      ],
      "projects": []
    },
    {
      "name": "IAM Members Viewer",
      "id": "iam-members-viewer",
      "type": "CHEF_MANAGED",
      "actions": [
        ...
      ],
      "projects": []
    },
    {
      "name": "role 1",
      "id": "r1",
      "type": "CUSTOM",
      "actions": [
        "do:the:thing"
      ],
      "projects": [
        "p1"
      ]
    },
    {
      "name": "Ingest",
      "id": "ingest",
      "type": "CHEF_MANAGED",
      "actions": [
        ...
      ],
      "projects": []
    },
    {
      "name": "Project Admin",
      "id": "project-admin",
      "type": "CHEF_MANAGED",
      "actions": [
        ...
      ],
      "projects": []
    },
    {
      "name": "Viewer",
      "id": "viewer",
      "type": "CHEF_MANAGED",
      "actions": [
        ...
      ],
      "projects": []
    }
  ]
}

$ sl | grep authz-service
[2019-04-23T22:52:19+00:00] authz-service.default(O): time="2019-04-23T22:52:19Z" level=info msg="Projects Authorized Query" action="iam:roles:list" projects="[]" resource="iam:roles" result="[p1 (unassigned)]" subject="[token:test]"
```

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [ ] Code actually executed?
- [ ] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/server/v2/authz.go [hunk 1]
```diff
@@ -68,6 +68,22 @@ func (s *authzServer) ProjectsAuthorized(
 		projects = req.ProjectsFilter
 	}
 
+	// this call checks for all allowed projects (without considering deny)
+	if version.Minor == api.Version_V1 && len(req.ProjectsFilter) == 0 {
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: I wonder if we're due for a helper that's something like `if beta2p1` or something . . . I know we *shouldn't* fall in here without `Major == v2` but still . . . this logic asks that question but makes the reader really have to 🤔 about it

File: components/authz-service/server/v2/authz.go [hunk 2]
```diff
@@ -82,18 +98,13 @@ func (s *authzServer) ProjectsAuthorized(
...
-			// an empty array, because an empty array coming back from the engine means none!
-			projectsAuthorized = []string{constants.AllProjectsExternalID}
-		} else {
-			// Engine allows all--but we want that to mean just the *requested* ones.
-			projectsAuthorized = req.ProjectsFilter
-		}
+	}
+
+	if stringutils.SliceContains(projectsAuthorized, constants.AllProjectsID) {
+		// Engine allows all and we requested all, so signify it as all.
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: This comment sounds very Gandalf-y, but it a bit confusing 😬

File: components/authz-service/server/v2/authz.go [hunk 3]
```diff
@@ -68,6 +70,27 @@ func (s *authzServer) ProjectsAuthorized(
...
+			Subjects: req.Subjects,
+			Pairs:    pairs,
+		}
+		resp, err := s.FilterAuthorizedProjects(ctx, &filterReq)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		projects = resp.Projects
+	}
```
Comment (tylercloke) [inline components/authz-service/server/v2/authz.go]: The logic here is getting mind-boggling (I know it was hard to follow to start with! Not blaming this change specifically). It might be better to just break this out into each case (v1.0, v2.0, v2.1 -- or at least v1.0 and v2.x) as separate functions instead of trying to handle them all with this branching. It's getting really, really hard to read through and understand the intentions behind the branching.

So you'd have 

```
func (s *authzServer) ProjectsAuthorized(...) {
         if version.Major == api.Version_V1 {
           projects = s.projectsAuthorizedV1(...)
        }
        ..similar helpers for V2.0 and V2.1

        // This looks like common clean up for every case and can probably stay.
	if stringutils.SliceContains(projectsAuthorized, constants.AllProjectsID) {
		// Engine allows all and we requested all, so signify it as all.
		// This must be different than the requested notion of all,
		// an empty array, because an empty array coming back from the engine means none!
		projectsAuthorized = []string{constants.AllProjectsExternalID}
	}

         return projects
}

func (s *authzServer) v1ProjectsAuthorized(...) {
         projects := []string{}
  	projectsAuthorized, err := s.engine.V2ProjectsAuthorized(ctx,
		engine.Subjects(req.Subjects),
		engine.Action(req.Action),
		engine.Resource(req.Resource),
		engine.ProjectList(projects...))
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())

       if len(projectsAuthorized) > 0 {
         projectsAuthorized = []string{constants.AllProjectsExternalID}
       }
       return projectsAuthorized;
}
```

I think it might be better to duplicate some of the code in `ProjectsAuthorized` at this point for the sake of readability and to make the intentions behind the different branch cases clear. 

**Disclaimer** The above code snippet is just to illustrate the idea and likely contains errors and misses opportunities to clean up the code more.

File: components/authz-service/server/v2/authz.go [hunk 4]
```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
...
+		} else {
+			// we pass the requested projects as is
+			projects = p
+		}
+	}
+
+	// while on v2, we always pass an empty project list to the engine because projects
+	// can't have an effect on the permission decision
+	if !isBeta2p1(v) {
+		// we override the requested projects because no filter should be applied on v2
```
Comment (msorens) [inline components/authz-service/server/v2/authz.go]: This comment essentially duplicates L270-271 so L273 could be safely deleted.

File: components/authz-service/server/v2/authz.go [hunk 5]
```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
...
+		projects = []string{}
+	}
+
+	return projects, nil
+}
+
+// parseAuthorizedProjects sets the engine's allowed project response based on the current IAM version
+func parseAuthorizedProjects(v api.Version, p []string) []string {
+	authorizedProjects := p
```
Comment (msorens) [inline components/authz-service/server/v2/authz.go]: Consider just naming the parameter in L281 `authorizedProjects` for simplicity.

File: components/authz-service/server/v2/authz_test.go [hunk 6]
```diff
@@ -78,10 +79,10 @@ func TestV2p1ProjectsAuthorized(t *testing.T) {
 				[]string{},    // no projects
 				[]string(nil), // well, almost verbatim
 			},
-			"when the request includes SOME projects and the engine response is ALL projects, returns the requested projects": {
+			"when the request includes SOME projects and the engine response is ALL projects, returns external ALL projects": {
 				[]string{"p1", "p2", "p3"},
 				[]string{constants.AllProjectsID, "p3"},
-				[]string{"p1", "p2", "p3"},
+				[]string{constants.AllProjectsExternalID},
```
Comment (msorens) [inline components/authz-service/server/v2/authz_test.go]: I was about to say this test result is wrong--it should not have changed.
But then I realized why. When I wrote this set of tests originally, I gave them the wrong set of inputs. They are testing _only_ what authz does with the OPA output. The requested projects should not be given as an input to these tests at all.
So if you don't mind cleaning these up while you're here, that would be useful.
(E.g. This test name should become "when then engine response is ALL projects, returns external ALL projects"

**Comment by msorens** [comment]
Oh, almost forgot. Probably should consider adding some tests to conformance_v2_test.go to cover the new functionality.

File: components/authz-service/engine/opa/policy/authz_v2_test.rego [hunk 7]
```diff
@@ -373,3 +373,56 @@ test_authorized_project_deny_real_data {
...
+				"sid-2": {"effect": "deny", "role": "operator", "resources": ["*"], "projects": [const_all_projects]},
+			},
+		}
+		 with input as {"subjects": ["x"], "action": "y", "resource": "z", "projects": ["p1", "p3", "p5"]}
+
+	actual_projects == set()
+}
+
+test_authorized_project_matches_only_allowed_projects_when_some_projects_denied {
+	actual_projects = authorized_project with data.roles.operator.actions as ["y"]
```
Comment (blakestier) [inline components/authz-service/engine/opa/policy/authz_v2_test.rego]: I'm still new to rego-- why use the `=` here instead of the `:=`? Are the tests functioning like rules? (Seems like `:=` is reserved for rules . . .)

File: components/authz-service/server/v2/authz.go [hunk 8]
```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
 	}()
 	return &x
 }
+
+func isBeta2p1(version api.Version) bool {
+	return version.Major == api.Version_V2 && version.Minor == api.Version_V1
+}
+
+// setRequestProjects sets the project filter based on the current IAM version
+func (s *authzServer) setRequestProjects(ctx context.Context, v api.Version, p []string) ([]string, error) {
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: What's `p` here? `projects`? `reqProjects`?

Comment (blakestier) [reply components/authz-service/server/v2/authz.go]: oops just saw @msorens's comment. +1

File: components/authz-service/server/v2/authz.go [hunk 9]
```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
...
+	return projects, nil
+}
+
+// parseAuthorizedProjects sets the engine's allowed project response based on the current IAM version
+func parseAuthorizedProjects(v api.Version, p []string) []string {
+	authorizedProjects := p
+	if !isBeta2p1(v) && len(p) > 0 {
+		// as long as at least one project is authorized
+		// we override the filtered projects with All Projects
+		// because no project filter should be applied on v2.0
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: I'd change this comment to `because no project filter should be applied unless on v2.1` to be consistent with other logic/naming.

File: components/authz-service/server/v2/authz.go [hunk 10]
```diff
@@ -235,3 +234,68 @@ func NewSwitch(c chan api.Version) *VersionSwitch {
...
+	if !isBeta2p1(v) {
+		// we override the requested projects because no filter should be applied on v2
+		projects = []string{}
+	}
+
+	return projects, nil
+}
+
+// parseAuthorizedProjects sets the engine's allowed project response based on the current IAM version
+func parseAuthorizedProjects(v api.Version, p []string) []string {
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: Two things:

1) The comment made me think we were setting the engine's response, but we're actually translating that response and setting the server's response, right? Can you clarify that?

2) Given 1, perhaps this method should be `translateAuthorizedProjects`? Or maybe something even longer to indicate that here we are translating OPA's response into a server response

File: api/interservice/authz/v2/policy.proto [hunk 11]
```diff
@@ -69,8 +69,15 @@ message Statement {
         items: { string: { pattern: "^[*]$|^[*]:[a-z][-a-zA-Z]*$|^[a-z][a-zA-Z]*:[a-z][a-zA-Z]*:[a-z][a-zA-Z]*$|^[a-z][a-zA-Z]*:[*]$|^[a-z][a-zA-Z]*:[*]:[a-z][a-zA-Z]*$|^[a-z][a-zA-Z]*:[a-z][a-zA-Z]*:[*]$" } } }];
 
     // references
-    string role = 4; // TODO: validations -- if we want a pattern, have it be `role:...`, otherwise, do other validations ;)
-    repeated string projects = 5; // same here
+    string role = 4 [(validate.rules).string.pattern = "$^|^[a-z0-9-]{1,64}$"];
```
[the remark this answers is not in the document]
Comment (bcmdarroch) [reply api/interservice/authz/v2/policy.proto]: oops 🤓

File: components/authz-service/engine/opa/policy/authz_v2_test.rego (hunk 7 above)
Comment (bcmdarroch) [reply components/authz-service/engine/opa/policy/authz_v2_test.rego]: good question. according to [opa's rego guide](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-write-policies/#the-basics), 
```
There’s one exception: if you use assignment := the compiler will check that the variable you are assigning has not already been used.

s {
    y := 41
    x := 42
    x > y
}
```

Comment (bcmdarroch) [reply components/authz-service/engine/opa/policy/authz_v2_test.rego]: and yeah, those cases are basically rego rules specifically written for testing. opa has [a good doc](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-test-policies/) on it

Comment (blakestier) [reply components/authz-service/engine/opa/policy/authz_v2_test.rego]: so we could use `:=` here?

File: components/authz-service/server/v2/authz.go (hunk 9 above)
Comment (bcmdarroch) [reply components/authz-service/server/v2/authz.go]: took this out

File: components/authz-service/server/v2/authz.go (hunk 3 above)
Comment (bcmdarroch) [reply components/authz-service/server/v2/authz.go]: agreed! about to push up some changes similar to what you suggested so i hope that helps!

File: components/authz-service/server/v2/authz.go (hunk 10 above)
Comment (bcmdarroch) [reply components/authz-service/server/v2/authz.go]: yeah i simplified this helper and renamed it

File: components/authz-service/server/v2/authz.go (hunk 2 above)
Comment (msorens) [reply components/authz-service/server/v2/authz.go]: Got a good chuckle from your comment here @blakestier !

File: components/authz-service/engine/conformance/conformance_v2_test.go [hunk 12]
```diff
@@ -262,10 +263,122 @@ func TestV2ProjectsAuthorized(t *testing.T) {
...
+						},
+					},
+				}
+				setPoliciesV2(t, e, pol)
+				actual, err := e.V2ProjectsAuthorized(args([]string{proj1, proj2}))
+				require.NoError(t, err)
+				assert.ElementsMatch(t, []string{}, actual)
+			})
+
+			t.Run("policy that denies all projects returns no projects when request contains no projects", func(t *testing.T) {
```
Comment (msorens) [inline components/authz-service/engine/conformance/conformance_v2_test.go]: Similar to above comment, we'd be better served by changing the name:
s/when request contains no projects/when all projects requested/

File: components/authz-service/engine/conformance/conformance_v2_test.go [hunk 13]
```diff
@@ -262,10 +263,122 @@ func TestV2ProjectsAuthorized(t *testing.T) {
...
+						"statement-id-1": map[string]interface{}{
+							"actions":   []string{act},
+							"resources": []string{res},
+							"effect":    "allow",
+							"projects":  []string{proj1},
+						},
+					},
+				}
+				setPoliciesV2(t, e, pol)
+				actual, err := e.V2ProjectsAuthorized(args([]string{}))
```
Comment (msorens) [inline components/authz-service/engine/conformance/conformance_v2_test.go]: And again use the new named constant `allProjectsRequest`

File: components/authz-service/engine/conformance/conformance_v2_test.go [hunk 14]
```diff
@@ -262,10 +263,122 @@ func TestV2ProjectsAuthorized(t *testing.T) {
 					},
 				}
 				setPoliciesV2(t, e, pol)
-				actual, err := e.V2ProjectsAuthorized(args())
+				actual, err := e.V2ProjectsAuthorized(args([]string{proj1, proj2}))
 				require.NoError(t, err)
 				assert.ElementsMatch(t, []string{proj2}, actual)
 			})
+
```
Comment (msorens) [inline components/authz-service/engine/conformance/conformance_v2_test.go]: You did a great job translating requirements into a new set of tests here, Brenna!

File: components/authz-service/server/v2/authz.go [hunk 15]
```diff
@@ -56,51 +58,53 @@ func (s *authzServer) IsAuthorized(
...
-			// an empty array, because an empty array coming back from the engine means none!
-			projectsAuthorized = []string{constants.AllProjectsExternalID}
-		} else {
-			// Engine allows all--but we want that to mean just the *requested* ones.
-			projectsAuthorized = req.ProjectsFilter
-		}
-	}
-
-	// TODO bd: parse result and send gateway different errors based on version
+	authorizedProjects = translateEngineResponse(engineResp)
```
[the remark this answers is not in the document]
Comment (bcmdarroch) [reply components/authz-service/server/v2/authz.go]: i think i'll leave it as is - it doesn't add more specificity
